### DISCUSSION TITLE
Stereology Measures

### DISCUSSION THREAD

**Discussion by amoebacodes** [neuropoly/axondeepseg#314] [Ideas of new features]
I have noticed that in the AxonDeepSeg paper, a grid is applied to small 512 x 512 pixels patches to overcome "border issues", meaning some axons land on the edges and thus are not properly analyzed. However, it is an issue still for axons on the edge of an entire image. 

Stereology measures could be utilized to circumvent this issue. As shown in the picture in the link below, a central area of the image can be carved out for analysis, whereas the crust can be used to assist analysis of axons that are on the border.
http://www.stereology.info/counting-rules/

Thank you for reading this and for considering this suggestion!

**Comment by mariehbourget** [comment]
Hi @amoebacodes,
Thanks for bringing that method up to our attention.

In AxonDeepSeg, we use an overlap to properly analyse axons landing on the edges of patches. That overlap acts as the crust from your example with stereology measures, and assist analysis on the border of the central area of patches.
For example, with a patch of 512x512 and an overlap of 25, the segmentation is performed on the entire patch of 512x512 for context, but only the 462x462 central area is output without the overlap. When all patches are analyzed, the 462x462 patches are then stitches together in the final segmentation (the overlap being used only for patches at the very edges of the image).

Let us know if you have more questions about patches behavior or if I misinterpreted your suggestion.

**Reply by akraskovSMDB** [reply]
Hi, somehow related question about how to treat/remove axons on the borders of the image.
I have independent images (no overlap or joint edges) and I want to remove from morphometric analysis all axons which are crossed/touched by edges. 
I can run through the morphometric output and check whether coordinates of the axon centre are closer to the edges than outer axon radius but this relies on axons being perfectly circular, which is rarely the case. I can be very conservative and remove all axons closer to the edges than diameter of the largest axon but will lose some good axons.
I was wondering whether there is an option to exclude from the morphometric analysis any axon which has a single pixel of the axon or myelin are on the border. 
Thank you

**Reply by hermancollin** [reply]
Hi @akraskovSMDB, thank you for reaching out.

Unfortunately, there isn't yet a way to remove axons from the morphometrics analysis, although we are working on a feature to filter the axons. This feature would let the user decide which axons he wants in his morphometrics analysis, so you could simply tell the program the index of the problematic axons and it would remove them.

Would it be possible to have a preview of the images you are working with? Maybe there would be an easier way to adress this problem.

**Comment by akraskovSMDB** [comment]
Hi, thank you for your answer.
I was more thinking about an additional column in the morphometric output file which will indicate whether a given axon is on the border or not. 
Like a couple of nice axons on the edge of this image
![image](https://user-images.githubusercontent.com/25688017/167168538-c74d544e-485f-446d-bb4f-73788dc4a464.png)

**Reply by hermancollin** [reply]
@akraskovSMDB Hmm I will check if this could be done easily and get back to you. Are you more worried about axons touching the border, myelin or both?

**Reply by akraskovSMDB** [reply]
Both would be great, either two additional columns or a code in one column (0 nothing on the border, 1 myelin on the border, 2 inner axon on the border)
thank you

**Reply by hermancollin** [reply]
@akraskovSMDB the quickest way to do this would be to save the bounding box of all elements in the morphometrics file, and then you could check if any of the bounds match the image border.

**Reply by akraskovSMDB** [reply]
Is it possible with a current version of axondeepseg_morphometrics? Thank you

**Reply by hermancollin** [reply]
@akraskovSMDB I will discuss this with the team and get back to you on Friday. There is an ongoing refactoring in the morphometrics code that could facilitate this.

**Reply by hermancollin** [reply]
@akraskovSMDB following an internal discussion, we decided to implement this feature. It should take 1-2 weeks to implement the solution, but I'll let you know as soon as we have a working prototype that you could use.

**Comment by hermancollin** [comment]
@akraskovSMDB an additionnal column specifying if the axonmyelin object is touching the image border is now available on the master branch, which you can update using `git pull; pip install -e .` inside the axondeepseg root folder.

The columns containing the axonmyelin's bounding box coordinates is a feature that should be ready next week.
